src/app/ReadHandler.h

```diff
@@ -0,0 +1,166 @@
+     *  @param[in]    apDelegate       InteractionModelDelegate set by application.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate);
+
+    /**
+     *  Shutdown the ReadHandler. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     */
+    void Shutdown();
+
+    /**
+     *  Reset the ReadHandler to initialized state
+     *
+     */
+    void Reset();
+
+    virtual ~ReadHandler() = default;
+
+    bool IsFree() const { return (nullptr == mpExchangeCtx); };
+
+    void OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                           const PayloadHeader & aPayloadHeader, System::PacketBufferHandle aPayload);
+
+    reporting::PriorityLevel FindNextPriorityForTransfer(void);
+
+    /**
+     *  Logger would save last logged event number for each logger buffer into schedule event number array
+     *  Then LoggingManagement could fetch new event until the schedule number for each buffer.
+     *
+     *  @param[in]    aLogger    The LoggingManagement object.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE if logger is invalid
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR SetEventLogEndpoint(reporting::LoggingManagement & aLogger);
+
+    /**
+     *  Compare last event number with vended event number for each bufer, if equal, it is up to date
+     *  Otherwise, it is not update to data.
+     *
+     *  @param[in]    aLogger    The LoggingManagement object.
+     *
+     *  @retval #Indiate if circular event buffer is up to date or not
+     *
+     */
+    bool CheckEventUpToDate(reporting::LoggingManagement & inLogger);
+
+    CHIP_ERROR SendReportData(System::PacketBufferHandle && aPayload);
+
+    CHIP_ERROR ProcessReadRequest(System::PacketBufferHandle && aPayload);
+
+    void OnReportProcessingComplete();
+
+    bool IsReportable(void) { return (mState == HandlerState::Initialized); }
+    bool IsReporting(void) { return (mState == HandlerState::ReportDataSending); }
+
+    void MoveToState(const HandlerState aTargetState);
+    void ClearState();
+    const char * GetStateStr() const;
+    CHIP_ERROR ClearExistingExchangeContext();
+
+    // The event number of the last event for each priority level
+    chip::EventNumber mSelfVendedEvents[CHIP_NUM_PRIORITY_LEVEL];
+
+    // The last schedule event number snapshoted in the beginning of reporting engine run
+    chip::EventNumber mLastScheduledEventNumber[CHIP_NUM_PRIORITY_LEVEL];
+
+    Messaging::ExchangeManager * mpExchangeMgr = nullptr;
+    Messaging::ExchangeContext * mpExchangeCtx = nullptr;
+    InteractionModelDelegate * mpDelegate      = nullptr;
+
+    //
```

Comment:
should a comment exit here? or just delete.

mCurrentPriority/mSuppressResponse/etc could have initializers like the pointers above.